### ISSUE TITLE
梯度异常 ，运行多轮后，grad_norm为nan

### ISSUE THREAD

**Issue by yanbozhang** [THUDM/GLM-4#174]
### System Info / 系統信息

{'loss': 0.0618, 'grad_norm': 0.58984375, 'learning_rate': 0.00026500000000000004, 'epoch': 5.27}                                                      
{'loss': 0.1924, 'grad_norm': 4.25, 'learning_rate': 0.000264, 'epoch': 5.29}                                                                          
{'loss': 0.1147, 'grad_norm': 1.34375, 'learning_rate': 0.000263, 'epoch': 5.31}                                                                       
{'loss': 0.1006, 'grad_norm': 5.625, 'learning_rate': 0.000262, 'epoch': 5.34}                                                                         
{'loss': 0.1459, 'grad_norm': 5.28125, 'learning_rate': 0.000261, 'epoch': 5.36}                                                                       
{'loss': 0.2096, 'grad_norm': nan, 'learning_rate': 0.00026000000000000003, 'epoch': 5.38} 

### Information / 问题信息

- [ ] The official example scripts / 官方的示例脚本
- [ ] My own modified scripts / 我自己修改的脚本和任务

### Reproduction / 复现过程

max_input_length: 1024
max_output_length: 512
training_args:
  # see `transformers.Seq2SeqTrainingArguments`
  output_dir: ./output
  max_steps: 5000
  # needed to be fit for the dataset
  learning_rate: 5e-4
  # settings for data loading
  per_device_train_batch_size: 1
  #gradient_accumulation_steps: 1
  dataloader_num_workers: 16
  remove_unused_columns: false
  # settings for saving checkpoints
  save_strategy: steps
  save_steps: 1000
  # settings for logging
  log_level: info
  logging_strategy: steps
  logging_steps: 10
  # settings for evaluation
  per_device_eval_batch_size: 4
  evaluation_strategy: steps
  eval_steps: 200
  # settings for optimizer
  # adam_epsilon: 1e-6
  # uncomment the following line to detect nan or inf values
  # debug: underflow_overflow
  predict_with_generate: true
  # see `transformers.GenerationConfig`
  generation_config:
    max_new_tokens: 256
  # set your absolute deepspeed path here
  # deepspeed: configs/ds_zero_2.json
peft_config:
  peft_type: LORA
  task_type: CAUSAL_LM
  r: 8
  lora_alpha: 32
  lora_dropout: 0.1

### Expected behavior / 期待表现

调整超参无效，环境与提供环境一致，有可能是模型forwad过程中某些linear层处理不当

**Comment by zRzRzRzRzRzRzR** [comment]
确是BF16微调的的吗

**Comment by yanbozhang** [comment]
> 确是BF16微调的的吗

fp16微调的

**Comment by zRzRzRzRzRzRzR** [comment]
fp16出现这个是正常的，溢出了，文档写了一定要bf16

**Comment by yanbozhang** [comment]
> fp16出现这个是正常的，溢出了，文档写了一定要bf16

如果要使用fp16，防止溢出，修改哪块代码可以实现呢

**Comment by zRzRzRzRzRzRzR** [comment]
> > fp16出现这个是正常的，溢出了，文档写了一定要bf16
> 
> 如果要使用fp16，防止溢出，修改哪块代码可以实现呢

这个问题应该是没办法解决吧，模型训练的问题
